### PR TITLE
Switch AI uses trapping abilities aggressively

### PR DESCRIPTION
**Pull request by Pawkkie**
## Description
Encourages the AI to use the trapping abilities Shadow Tag, Arena Trap, and Magnet Pull in its switching decisions. Post-KO trapping switches are tied to the AI_FLAG_SMART_MON_CHOICES, and mid-battle trapping switches are tied to AI_FLAG_SMART_SWITCHING (which requires AI_FLAG_SMART_MON_CHOICES anyways).

The AI will switch in a trapper under the following conditions:
- The player has other mons remaining in their party
- The AI has a mon with an ability that can trap the player's current battler
- The AI's trapping mon can win a 1v1 against the player's current mon

"Winning the 1v1" in this case is defined as KO'ing the player before they get KO'd themselves, based on the same hitsToKO calculations used elsewhere in the switch and scoring AI. These thresholds are different depending on whether the AI is switching mid-battle and needs to tank a hit vs after a KO and doesn't need to, and both cases are accounted for.

I've also included tests for both mid-battle and post-KO cases, as well as one to ensure the trapping mon isn't switched in preferentially when the player doesn't have any mons in their party.

I'm open to the idea of tying some random roll to these behaviours to make them more difficult to predict, though it will also make them more difficult to test as far as I understand it. The HasBadOdds() function uses a 50% chance of failure as a benchmark, but that function is significantly more likely to lead to switch spam than the trapper conditions, which won't lead to switch spam and will instead just deterministically switch in the trapper when the conditions are met. I'm not sure what threshold would be considered appropriate if any at all, but I'm happy to add one in to increase the variance if that's preferred! 

This also includes some very minor refactoring in `GetBestMonIntegrated` to clean up some conditionals for readability.

## **Discord contact info**
@Pawkkie

### CODE REVIEW

File: src/battle_ai_switch_items.c [hunk 1]
```diff
@@ -1691,6 +1736,25 @@ static s32 GetMaxDamagePlayerCouldDealToSwitchin(u32 battler, u32 opposingBattle
     return maxDamageTaken;
 }
 
+static bool32 CanAbilityTrapOpponent(u16 ability, u32 opponent)
+{
+    if ((B_GHOSTS_ESCAPE >= GEN_6 && IS_BATTLER_OF_TYPE(opponent, TYPE_GHOST)))
+        return FALSE;
+    else if (ability == ABILITY_SHADOW_TAG)
+    {
+        if (B_SHADOW_TAG_ESCAPE >= GEN_4 && GetBattlerAbility(opponent) == ABILITY_SHADOW_TAG) // Check if ability exists in species
```
Comment (AlexOn1ine) [inline src/battle_ai_switch_items.c:1745]: Why does the ability need to be checked again? If you pass the ability into this function isn't it proof that the mon has it?

File: src/battle_ai_switch_items.c [hunk 2]
```diff
@@ -1851,13 +1916,26 @@ static u32 GetBestMonIntegrated(struct Pokemon *party, int firstId, int lastId,
...
                             slowThreatenId = i;
                         }
                     }
                 }
+
+                // If mon can trap
+                if (CanAbilityTrapOpponent(AI_DATA->switchinCandidate.battleMon.ability, opposingBattler))
+                {
+                    hitsToKOPlayer = GetNoOfHitsToKOBattlerDmg(damageDealt, opposingBattler);
+                    // Good grief clean up the 1v1 check and outspeed check
```
Comment (AlexOn1ine) [inline src/battle_ai_switch_items.c:1931]: The comment should probably be removed.

File: src/battle_ai_switch_items.c [hunk 3]
```diff
@@ -1851,13 +1916,26 @@ static u32 GetBestMonIntegrated(struct Pokemon *party, int firstId, int lastId,
...
+                    if (CountUsablePartyMons(opposingBattler) > 0
+                        && (((hitsToKOAI > hitsToKOPlayer && isSwitchAfterKO) // If can 1v1 after a KO
+                        || (hitsToKOAI == hitsToKOPlayer && isSwitchAfterKO && (aiMonSpeed > playerMonSpeed || aiMovePriority > 0)))
+                        || ((hitsToKOAI > hitsToKOPlayer + 1 && !isSwitchAfterKO) // If can 1v1 after mid battle
+                        || (hitsToKOAI == hitsToKOPlayer + 1 && !isSwitchAfterKO && (aiMonSpeed > playerMonSpeed || aiMovePriority > 0)))))
```
Comment (AlexOn1ine) [inline src/battle_ai_switch_items.c:1936]:
```suggestion
            if (CountUsablePartyMons(opposingBattler) > 0
            && (((hitsToKOAI > hitsToKOPlayer && isSwitchAfterKO) // If can 1v1 after a KO
            || (hitsToKOAI == hitsToKOPlayer && isSwitchAfterKO && (aiMonSpeed > playerMonSpeed || aiMovePriority > 0)))
            || ((hitsToKOAI > hitsToKOPlayer + 1 && !isSwitchAfterKO) // If can 1v1 after mid battle
            || (hitsToKOAI == hitsToKOPlayer + 1 && !isSwitchAfterKO && (aiMonSpeed > playerMonSpeed || aiMovePriority > 0)))))
```

File: src/battle_ai_switch_items.c [hunk 4]
```diff
@@ -1867,8 +1945,11 @@ static u32 GetBestMonIntegrated(struct Pokemon *party, int firstId, int lastId,
...
     if (isSwitchAfterKO)
     {
-        // Return GetBestMonRevengeKiller >  GetBestMonTypeMatchup > GetBestMonBatonPass > GetBestMonDmg
-        if (revengeKillerId != PARTY_SIZE)
+        // Return Trapper > GetBestMonRevengeKiller >  GetBestMonTypeMatchup > GetBestMonBatonPass > GetBestMonDmg
+        if (trapperId != PARTY_SIZE)
+            return trapperId;
+
+        else if (revengeKillerId != PARTY_SIZE)
             return revengeKillerId;
```
Comment (AlexOn1ine) [inline src/battle_ai_switch_items.c:1953]: I would remove all the blank lines it this if-else block

**Review by AlexOn1ine — COMMENTED** [review]
Few things but looks good otherwise

File: src/battle_ai_switch_items.c (hunk 1 above)
Comment (Pawkkie) [reply src/battle_ai_switch_items.c:1745]: I might be misunderstanding, abilities aren't checked before being passed to `CanAbilityTrapOpponent`, and Shadow Tag specifically needs to be checked on both the AI and the player to see if they cancel each other out.

I think I'm missing what you're talking about specifically, would you mind clarifying? :)

File: src/battle_ai_switch_items.c (hunk 2 above)
Comment (Pawkkie) [reply src/battle_ai_switch_items.c:1931]: LOL yeah good catch that was a note-to-self during cleanup

File: src/battle_ai_switch_items.c (hunk 1 above)
Comment (AlexOn1ine) [reply src/battle_ai_switch_items.c:1745]: oh nevermind. I've only now realized that it is checking opponent. I forgot this ability effect exists.

Comment (Pawkkie) [reply src/battle_ai_switch_items.c:1745]: No worries! :D

**Comment by Pawkkie** [comment]
Ready for re-review :)

File: test/battle/ai.c [hunk 5]
```diff
@@ -955,3 +955,62 @@ AI_DOUBLE_BATTLE_TEST("AI will the see a corresponding absorbing ability on part
...
+}
+
+AI_SINGLE_BATTLE_TEST("AI_FLAG_SMART_MON_CHOICES: AI will switch in trapping mon after KO")
+{
+    u32 aiSmartMonChoicesFlag = 0; // Enables trapping behaviour after KOs
+    PARAMETRIZE { aiSmartMonChoicesFlag = 0; } // No trapping behaviour
+    PARAMETRIZE { aiSmartMonChoicesFlag = AI_FLAG_SMART_MON_CHOICES; } // Traps with mid battle switches
+    GIVEN{
+        AI_FLAGS(AI_FLAG_CHECK_BAD_MOVE | AI_FLAG_CHECK_VIABILITY | AI_FLAG_TRY_TO_FAINT | aiSmartMonChoicesFlag);
+        PLAYER(SPECIES_MAWILE) { Speed(2); Moves(MOVE_PROTECT, MOVE_TACKLE); }
```
Comment (AlexOn1ine) [inline test/battle/ai.c:989]: Could you add the assume for Steel Type Mawile?

File: test/battle/ai.c [hunk 6]
```diff
@@ -955,3 +955,62 @@ AI_DOUBLE_BATTLE_TEST("AI will the see a corresponding absorbing ability on part
...
+}
+
+AI_SINGLE_BATTLE_TEST("AI won't use trapping behaviour if player only has 1 mon left")
+{
+    u32 aiSmartMonChoicesFlag = 0; // Enables trapping behaviour after KOs
+    PARAMETRIZE { aiSmartMonChoicesFlag = 0; } // No trapping behaviour
+    PARAMETRIZE { aiSmartMonChoicesFlag = AI_FLAG_SMART_MON_CHOICES; } // Traps with mid battle switches
+    GIVEN{
+        AI_FLAGS(AI_FLAG_CHECK_BAD_MOVE | AI_FLAG_CHECK_VIABILITY | AI_FLAG_TRY_TO_FAINT | aiSmartMonChoicesFlag);
+        PLAYER(SPECIES_MAWILE) { Speed(2); Moves(MOVE_PROTECT, MOVE_TACKLE); }
```
Comment (AlexOn1ine) [inline test/battle/ai.c:1009]: here as well

**Review by AlexOn1ine — COMMENTED** [review]
Sorry didn't catch those first. 

Non blocking but do you think it makes sense to make a new test file for the switch ai tests?

**Comment by Pawkkie** [comment]
> Non blocking but do you think it makes sense to make a new test file for the switch ai tests?

YES

I thought you'd never ask <3

I'll move everything that has a switch AI flag into its own file called ai_switching.c :)

**Comment by Pawkkie** [comment]
> Non blocking but do you think it makes sense to make a new test file for the switch ai tests?

This was an extra good idea because I found
```
AI_SINGLE_BATTLE_TEST("AI gets baited by Protect Switch tactics") // This behavior is to be fixed.
```
while doing it, which links all the way back to #3346 . D'you think we should wait for Egg to be around or is this something we can brainstorm for and then ping Egg when we have an implementation idea? It's something I'm interesting in poking around with in the near future now that I've remembered it exists :)

**Comment by AlexOn1ine** [comment]
> while doing it, which links all the way back to #3346 . D'you think we should wait for Egg to be around or is this something we can brainstorm for and then ping Egg when we have an implementation idea? It's something I'm interesting in poking around with in the near future now that I've remembered it exists :)

You could ping him if you want but I don't think we should intentionally wait for him.

**Review by AlexOn1ine — APPROVED** [review]

**Pull request merged**